src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1916 @@
+            _captureNumbersToSpan = captureNumbersToSpan;
+
+            // Get the first token.  It is allowed to have trivia on it.
+            ConsumeCurrentToken(allowTrivia: true);
+        }
+
+        /// <summary>
+        /// Returns the latest token the lexer has produced, and then asks the lexer to 
+        /// produce the next token after that.
+        /// </summary>
+        /// <param name="allowTrivia">Whether or not trivia is allowed on the next token
+        /// produced.  In the .net parser trivia is only allowed on a few constructs,
+        /// and our parser mimics that behavior.  Note that even if trivia is allowed,
+        /// the type of trivia that can be scanned depends on the current RegexOptions.
+        /// For example, if <see cref="RegexOptions.IgnorePatternWhitespace"/> is currently
+        /// enabled, then '#...' comments are allowed.  Otherwise, only '(?#...)' comemnts
+        /// are allowed.</param>
+        private RegexToken ConsumeCurrentToken(bool allowTrivia)
+        {
+            var previous = _currentToken;
+            _currentToken = _lexer.ScanNextToken(allowTrivia, _options);
+            return previous;
+        }
+
+        /// <summary>
+        /// Given an input text, and set of options, parses out a fully representative syntax tree 
+        /// and list of diagnotics.  Parsing should always succeed, except in the case of the stack 
+        /// overflowing.
+        /// </summary>
+        public static RegexTree TryParse(ImmutableArray<VirtualChar> text, RegexOptions options)
+        {
+            try
+            {
+                // Parse the tree once, to figure out the capture groups.  These are needed
+                // to then parse the tree again, as the captures will affect how we interpret
+                // certain things (i.e. escape references) and what errors will be reported.
+                //
+                // This is necessary as .net regexes allow references to *future* captures.
+                // As such, we don't know when we're seeing a reference if it's to something
+                // that exists or not.
+                var tree1 = new RegexParser(text, options,
+                    ImmutableDictionary<string, TextSpan>.Empty,
+                    ImmutableDictionary<int, TextSpan>.Empty).ParseTree();
+
+                var analyzer = new CaptureInfoAnalyzer(text);
+                var (captureNames, captureNumbers) = analyzer.Analyze(tree1.Root, options);
+
+                var tree2 = new RegexParser(
+                    text, options, captureNames, captureNumbers).ParseTree();
+                return tree2;
+            }
+            catch (Exception e) when (StackGuard.IsInsufficientExecutionStackException(e))
+            {
+                return null;
+            }
+        }
+
+        private RegexTree ParseTree()
+        {
+            var expression = this.ParseAlternatingSequences(consumeCloseParen: true);
```

Comment:
Why consume close paren? We didn't consume any open paren yet.

Reply:
I'll add comments.  The idea here is that the parent calls to say if they were the ones that consumed the open paren (and thus want to be the ones comsuming the close paren).  In this case, as the topmost call, we are not parsing out the open paren, and thus we would like the ParseAlternatingSequences call to consume all the excess close parens that are are seen and not bail out when encountering them.

Reply:
#resolved.